lib/ecto/exceptions.ex

```diff
@@ -81,22 +81,45 @@ defmodule Ecto.InvalidChangesetError do
   defexception [:action, :changeset]
 
   def message(%{action: action, changeset: changeset}) do
+    changes = extract_changes(changeset)
+    errors = Ecto.Changeset.traverse_errors(changeset, & &1)
+
     """
     could not perform #{action} because changeset is invalid.
 
-    * Changeset changes
+    Changes
+
+    #{pretty changes}
 
-    #{inspect changeset.changes}
+    Params
 
-    * Changeset params
+    #{pretty changeset.params}
 
-    #{inspect changeset.params}
+    Errors
 
-    * Changeset errors
+    #{pretty errors}
 
-    #{inspect changeset.errors}
+    Changeset
+
+    #{pretty changeset}
     """
   end
+
+  defp pretty(term) do
+    inspect(term, pretty: true)
+    |> String.split("\n")
+    |> Enum.map_join("\n", &"    " <> &1)
+  end
+
+  defp extract_changes(%Ecto.Changeset{changes: changes}) do
```

Comment:
@josevalim do you mean the "main" changeset is marked as delete or one of the embeds changesets? If the main changeset is marked as default, do you mean that we wouldn't show the last part:

```
     ** (Ecto.InvalidChangesetError) could not perform delete because changeset is invalid.

     Applied changes

         %{posts: [nil, %{}, %{title: "new"}, %{title: nil}, %{title: "new name"}]}

     Params

         %{"posts" => [%{"title" => "new"}, %{"id" => "2", "title" => nil},
            %{"id" => "3", "title" => "new name"}]}

     Errors

         %{posts: [%{}, %{}, %{},
            %{title: [{"can't be blank", [validation: :required]}]}, %{}]}

     (Changeset used to be here)
```

or something else?